### ISSUE TITLE
problem: users can't fix store names

### ISSUE THREAD

**Issue by olliechick** [howlongistheline/howlongistheline.org#133]
When store names are incorrect/incomplete, there is no way for users to fix them or report them as wrong, other than contacting someone, e.g. through the Facebook group or the GitHub issues.

The page that is used to report duplicate stores could also be used to report incorrect store names (and possibly addresses, too).

**Comment by jimkleine** [comment]
@olliechick @gazhayes I did complete a fairly substantial clean-up of the store names and addresses on Friday night.  The updates aren't in production yet.  That largely resolved e.g. #117, but the longer we leave the merge, the more fraught this becomes.

There is a fundamental problem with allowing unconstrained edits, which relates to stores with ambiguous names and locations.  #117 identified that the Countdown Highland Park store is not actually in the Highland Park shopping centre: that store is actually Countdown Aviemore Drive.

**Comment by LyallD** [comment]
Great site, but some stores are so ambiguous that the information is meaningless. For example, Countdown Hamilton - like which of the 8 or so Countdown stores in Hamilton is that?  A street address or suburb needs to be mandatory (logged as a separate issue).

**Comment by olliechick** [comment]
@jimkleine has your cleanup been pushed into production yet?

**Comment by jimkleine** [comment]
@olliechick Unfortunately not: I don't have access to the production DB. @gazhayes Did you get the merge script I sent you and will you have a chance to look at this tonight?

**Comment by gazhayes** [comment]
Hi @olliechick and @jimkleine, sorry, I will definitely get to that in the morning! I've been catching up on other work over the last few days but will be able to get back into this project tomorrow :)

**Comment by gazhayes** [comment]
Ok the database has been updated. Safer and more sustainable to avoid that in future and just work at a 'protocol' level to allow users to fix things themselves - if there's a problem with the data the real problem is with the site itself.

**Comment by jimkleine** [comment]
@gazhayes (1) I'm not yet seeing the name/address updates. Did you restore the "groomed" collection first?

(2) This was an organic problem. It arose from crowd sourcing the locations without first providing clear guidance about/validation for what constitutes a "good" store name and address.

(3)  Since locations are effectively required to be added on mobile devices, we need to be supportive of common typos, incorrect capitalisation etc.

**Comment by gazhayes** [comment]
Strange, yeah I restored the groomed data and ran the update. You're not seeing *any* of them, right?

The production environment isn't actually running Meteor (it runs with Node instead) and I'm using mongo replicaset. This is how I pushed the update:

```
mongodump --host HLITL-shard-0/hlitl-shard-00-00-cowgv.mongodb.net:27017,hlitl-shard-00-01-cowgv.mongodb.net:27017,hlitl-shard-00-02-cowgv.mongodb.net:27017 --ssl --username howlongistheline --password xxxx --authenticationDatabase admin --db production
tar -zcvf productiondump.tar.gz dump
mv dump/production dump/meteor
mongorestore -h 127.0.0.1 --port 3001 -d meteor dump/meteor 
mongorestore -h 127.0.0.1 --port 3001 -d meteor ../dump/meteor #groomed data
mongo localhost:3001/meteor NameAddressUpdate.js
rm -rf dump
mongodump -h 127.0.0.1 --port 3001 -d meteor
mv dump/meteor dump/production
mongorestore --host HLITL-shard-0/hlitl-shard-00-00-cowgv.mongodb.net:27017,hlitl-shard-00-01-cowgv.mongodb.net:27017,hlitl-shard-00-02-cowgv.mongodb.net:27017 --ssl --username howlongistheline --password xxxx --authenticationDatabase admin 

```

**Comment by jimkleine** [comment]
@gazhayes I don't believe you have any of the updates (try using "Countdown" as a search criteria and it returns lots of store names that are just "Countdown", which I groomed out).

1.  The script sets its own connection string, although I it should have defaulted to the string you passed to the mongo shell anyway.

2.  I read this as you backed up the production database, restored it to a separate instance, applied the script, backed up the separate instance and then restored to production.  I think there might be a [logical flaw in that process](https://docs.mongodb.com/manual/reference/program/mongorestore/#bin.mongorestore):

> mongorestore can create a new database or add data to an existing database. However, mongorestore performs inserts only and does not perform updates. That is, if restoring documents to an existing database and collection and existing documents have the same value _id field as the to-be-restored documents, mongorestore will not overwrite those documents.

You can use the --drop argument to mongorestore, but that is likely to cause a service outage.

3.  My background is large relational database systems, but my first thought is that sharding a database this small is likely to be a performance negative.  The collections are sufficiently small that they should be fully cached in memory, which implies nanosecond-scale response times.  Network latency is millisecond-scale at best.  I'll qualify this by saying that I need to investigate how Mongo manages transactions (updates) across the replicas and without knowing whether the number of concurrent network connections to the DB host is currently a bottleneck (in an ideal world, the web hosts would be using connection pools).

**Comment by gazhayes** [comment]
Just quickly, the production database was dropped before restoring, we were actually running on 'test' not 'production' before this but I changed the wording in the script I posted to make it flow better.

The reason for using a replicaset is actually just for access redundancy, there are now 8 application servers doing simultaneous reads.

I emailed you the credentials, don't break anything!

Also, I don't see any countdown stores that are just 'countdown' with no address?

**Comment by olliechick** [comment]
A good example, as someone has pointed out before (I think on the Facebook
page), is if you search "hamilton" the second result is just called
"Countdown" and the address is just "Hamilton" - there is 8 or so
Countdowns in Hamilton

**Comment by gazhayes** [comment]
Ah ok fair enough then

**Comment by jimkleine** [comment]
@gazhayes Got your e-mail.  Credentials aren't working for me (I think this is a firewall issue on your end).  Please check your e-mail.

**Comment by jimkleine** [comment]
@gazhayes I completed a check of the database tonight: check your e-mail for details.  There are around ~300 new stores that will also require their names and addresses checked.  I'll try to get to that over the next 24 hours.

**Comment by jimkleine** [comment]
@gazhayes We still seem to be waiting on you to correct the configuration on the web nodes re MONGO_URL.  Do you have an ETA on this please, because this is on the critical path to get the store data fixes live?

**Comment by gazhayes** [comment]
@jimkleine Ah yeah, I'll re-run those scripts and update the mongourl in the morning.

**Comment by jimkleine** [comment]
@gazhayes I'm still not seeing the store name and address clean ups.  I'm likely to re-apply these later today unless there is a reason I shouldn't?

**Comment by gazhayes** [comment]
@jimkleine no reason not to.

I really don't know why it didn't work :(

Everything is running on the 'live' db, so just take a backup and you could run it straight on that so that no live data gets lost during dump/restore.

**Comment by jimkleine** [comment]
@gazhayes Is there a lull in traffic at any time during a typical day that provides a good window for this (The backup will be the biggest server load, but I don't want to inadvertently choose "rush hour").

**Comment by gazhayes** [comment]
Not really, other than after 9pm, but this is a relatively fast operation so it should be fine.

The only problem is if you do a dump and restore, in that case the most important thing is having it in a script so that there's minimum time between them for people to input and lose data

**Comment by jimkleine** [comment]
@gazhayes My updates are in as of ~6:15pm.  I'm going to try to make a dent in cleaning up the newer stores overnight.  I have a local backup that I took just prior to that should you need it for any reason.

**Comment by jimkleine** [comment]
We are currently adding ~5 stores/day and three of those are typically duplicates.  I'm only spending ~5 mins/day confirming addresses (two of today's additions were "Countdown stores" that mapped to private homes) and purging the duplicates, so I'm going to close this for now.  We can revisit it if the volume changes.
